File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pro = 1
        dic = {0:0}
        for i in nums:
            if i == 0:
                dic[0] += 1
            else:
                pro = pro * i
        print(pro)
        for j in range(len(nums)):
            if dic[0] == 0:
                nums[j] = pro // nums[j]
            elif dic[0] == 1:
                if nums[j] == 0:
                    nums[j] = pro
                else:
                    nums[j] = 0
            else:
                nums[j] = 0
        return nums

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefix(n, 1);
        vector<int> suffix(n, 1);
        for(int i = 1; i < nums.size(); i++) {
            prefix[i] = prefix[i - 1] * nums[i - 1];
        }
        for(int j = nums.size() - 2; j >= 0 ; j--) {
            suffix[j] = suffix[j + 1] * nums[j + 1];
        }
        for(int i = 0; i < nums.size(); i++) {
            nums[i] = prefix[i] * suffix[i];
        }
        return nums;
    }
};